File: src/main/java/cn/afeibaili/mchat/socket/Reader.java
package cn.afeibaili.mchat.socket;

import cn.afeibaili.mchat.socket.cipher.CipherProcessor;

import java.io.BufferedReader;
import java.io.Closeable;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.Socket;
import java.util.function.Consumer;

/**
 * 读取器类
 *
 * @author AfeiBaili
 * @version 2026/3/3 00:26
 */

public class Reader implements Closeable {
    Socket socket;
    CipherProcessor cipher;
    Consumer<String> action;
    Consumer<Exception> cat;

    Thread readThread;
    BufferedReader reader;

    public Reader(Socket socket, CipherProcessor cipher, Consumer<String> action, Consumer<Exception> cat) {
        this.socket = socket;
        this.cipher = cipher;
        this.action = action;
        this.cat = cat;

        try {
            reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        readThread = new Thread(() -> {
            try {
                String line;
                while (!Thread.currentThread().isInterrupted() && (line = reader.readLine()) != null) {
                    action.accept(cipher.decrypt(line));
                }
                throw new RuntimeException("MChat server disconnect");
            } catch (Exception e) {
                e.printStackTrace();
                cat.accept(e);
            }
        });
        readThread.start();
    }

    @Override
    public void close() {
        try {
            reader.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        readThread.interrupt();
    }
}

File: core/src/main/kotlin/online/afeibaili/mchat/socket/Reader.kt
package online.afeibaili.mchat.socket

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import online.afeibaili.mchat.logger.Logger
import online.afeibaili.mchat.socket.cipher.CipherProcessor
import java.io.Closeable
import java.net.Socket

/**
 * 读取器
 *
 *@author AfeiBaili
 *@version 2025/11/3 19:11
 */

class Reader(
    socket: Socket,
    val cipher: CipherProcessor,
    val action: (String) -> Unit,
    val catch: (Throwable) -> Unit,
) : Closeable {
    private val scope = CoroutineScope(Dispatchers.IO)
    private val reader = socket.inputStream.bufferedReader()
    private val logger = Logger.getLogger("MChatReader")
    var isClosed: Boolean = false

    init {
        scope.launch {
            runCatching {
                var line: String
                while (reader.readLine().also { line = it } != null) {
                    val message: String = cipher.decrypt(line)
                    action(message)
                    logger.info(message)
                }
                throw RuntimeException("MChat server disconnect")
            }.onFailure { exception ->
                if (!isClosed) catch(exception)
            }
        }
    }

    override fun close() {
        reader.close()
        scope.cancel()
    }
}